devito/types/caching.py

```diff
@@ -165,6 +172,17 @@ def clear(cls, force=True):
         else:
             gc.collect()
 
-        for key, obj in list(_SymbolCache.items()):
+        # mydict.copy() is safer than list(mydict) for getting an unchanging list
+        # See https://bugs.python.org/issue40327 for terrifying discussion
+        # on this issue.
+        # This is a separate line so that when we get a "dictionary keys changed
+        #  during iteration" error we know which operation caused it
+        cache_copied = _SymbolCache.copy()
+        for key in cache_copied:
+            obj = _SymbolCache.get(key)
+            if obj is None:
+                # deleted by another thread
+                continue
             if obj() is None:
-                del _SymbolCache[key]
+                # does not error if already gone
```

Comment:
either this comment, or the one above, you may want to mention that all of this is to support multi-threading

Reply:
done